### ISSUE TITLE
Any call to OSSL_ENCODER_CTX_new_byXXX needs to support passing a libctx.

### ISSUE THREAD

**Issue by slontis** [openssl/openssl#13542]
This is related to some new encoder code that ignores the libctx:

e.g:
PEM_write_bio_Parameters() uses the macro IMPLEMENT_PEM_provided_write_body_vars()
which then calls
OSSL_ENCODER_CTX_new_by_##type(x, PEM_SELECTION_##asn1, "PEM", PEM_STRUCTURE_##asn1,  NULL, NULL);
The last 2 params are the libctx, propq which will not work if they are just NULL..

So we either need to extract the libctx (which doesnt address that there is no propq)..
using 
````
const OSSL_PROVIDER *prov = EVP_KEYMGMT_provider(pkey->keymgmt);
OSSL_LIB_CTX *libctx = ossl_provider_libctx(prov);
````
OR have a PEM_write_bio_Parameters_ex(libctx, propq) ??

To reproduce this locally do..
````
export OPENSSL_TEST_LIBCTX=1
apps/openssl genpkey -genparam -algorithm DSA
````

**Comment by levitte** [comment]
> So we either need to extract the libctx (which doesnt address that there is no propq) using 
> ````
> const OSSL_PROVIDER *prov = EVP_KEYMGMT_provider(pkey->keymgmt);
> OSSL_LIB_CTX *libctx = ossl_provider_libctx(prov);
> ````

Side note: this actually had me wonder why EVP_ENCODER_CTX_new_by_XXX() should even take a libctx argument.  I had that because OSSL_DECODER_CTX_new_by_XXX does, but while the latter *does* need it, the former...  doesn't really.

**Comment by slontis** [comment]
ossl_encoder_ctx_setup_for_EVP_PKEY() calls OSSL_ENCODER_CTX_get_num_encoders() which returns 0..
i.e: it cant find the encoders as they are not in the default lib context

**Comment by levitte** [comment]
Was that an answer to my side note?

**Comment by levitte** [comment]
Maybe I didn't make myself clear enough.  The question in my mind is, why doesn't `EVP_ENCODER_CTX_new_by_PKEY()` do this internally instead of being passed a libctx?

``` C
const OSSL_PROVIDER *prov = EVP_KEYMGMT_provider(pkey->keymgmt);
OSSL_LIB_CTX *libctx = ossl_provider_libctx(prov);
```

In other words, instead of passing an explicit libctx to `EVP_ENCODER_CTX_new_by_XXX()`, let it extract it from the object it's supposed to encode?

**Comment by levitte** [comment]
Either way, for this issue, I [ahem] have a branch lying around.  It's half ready, just needs some finishing touches.

**Comment by slontis** [comment]
> In other words, instead of passing an explicit libctx to `EVP_ENCODER_CTX_new_by_XXX()`, let it extract it from the object it's supposed to encode?

And my subquestion was where does the propq come from..

**Comment by levitte** [comment]
> And my subquestion was where does the propq come from..

I didn't say anything about propq.  That one still needs to be passed explicitly.

**Comment by slontis** [comment]
> I didn't say anything about propq. That one still needs to be passed explicitly.

Regardless of which parameter it is - it is not being propagated. So we probably still need the _ex variant anyway..

**Comment by levitte** [comment]
> Regardless of which parameter it is - it is not being propagated. So we probably still need the _ex variant anyway..

I never questioned that.  My side note was a *side note*, not a direct answer to this issue.

**Comment by levitte** [comment]
I'll raise another issue regarding that side note

**Comment by levitte** [comment]
Proposed fix in #13547
